### DISCUSSION TITLE
Try to connect the strongswan ipsec vpn over dhcp

### DISCUSSION THREAD

**Discussion by shankar25121995** [strongswan/strongswan#2291] [General]
i have windows fortinetvpn client settings for reference

connection name : vakrangee1234
Remote gateway:157.119.230.26
authentication method : pre-shared key
pre-shared key = V0rtex@123
Xauth : save login
username: TLS_VORTEX
password : tL$8907
vpn settings:

IKE
mode
**options: DHCP over IPsec**
enable IPv4 Split Tunnel
10.101.0.0 255.255.0.0

phase1:

IKE proposal : AES128-AES256-sha256
DH group:5
key life :86400sec
dead peer detection
NAT traversal

Phase 2:

IKE proposal
keylife:43200sec
enable replay detection
enable perfect forward secrecy(PFS)
DH group:5
this configuration works well in windows
i tried the same settings in debian 10 with strongswan ipsec
ipsec.conf configuration settings for reference

config setup
    charondebug="ike 2, knl 2, cfg 2, net 2, esp 2, dmn 2, 0"
    uniqueids=no

conn vakrangee1234
    keyexchange=ikev1
    aggressive=no
    type=tunnel
    authby=xauthpsk
    xauth=client
    leftauth=psk
    leftauth2=xauth
    xauth_identity=TLS_VORTEX
    left=%defaultroute
    leftsubnet=0.0.0.0/0
    leftsourceip=%config
    right=157.119.230.26
    rightsubnet=10.101.0.0/16
    rightid=%any
    rightauth=psk
    ike=aes128-aes256-sha256-modp1536
    ikelifetime=86400s
    esp=aes128-aes256-sha256-modp1536
    lifetime=43200s
    reauth=yes
    keyingtries=%forever
    dpdaction=hold
    modeconfig=pull
    forceencaps=yes
    auto=start
**i tried this configuration and so many changes done in this config still connection not esablished,in some times connection will be establish but virtual ip not assainging,kindly support on this i was stuck on this because most of the client machines is debian only.**
![Screenshot from 2024-06-16 21-43-32](https://github.com/strongswan/strongswan/assets/63556353/4028e04b-b939-44b5-9b62-d7f1298d587f)

**syslog for reference** 

Jun 16 21:42:39 debian charon: 00[DMN] Starting IKE charon daemon (strongSwan 5.7.2, Linux 4.19.0-26-686-pae, i686)
Jun 16 21:42:39 debian charon: 00[CFG] PKCS11 module '<name>' lacks library path
Jun 16 21:42:39 debian charon: 00[LIB] Padlock features supported:, enabled:
Jun 16 21:42:39 debian charon: 00[KNL] known interfaces and IP addresses:
Jun 16 21:42:39 debian charon: 00[KNL]   lo
Jun 16 21:42:39 debian charon: 00[KNL]     127.0.0.1
Jun 16 21:42:39 debian charon: 00[KNL]     ::1
Jun 16 21:42:39 debian charon: 00[KNL]   ens33
Jun 16 21:42:39 debian charon: 00[KNL]     192.168.0.2
Jun 16 21:42:39 debian charon: 00[KNL]     fe80::20c:29ff:fe09:38ab
Jun 16 21:42:39 debian charon: 00[CFG] loading ca certificates from '/etc/ipsec.d/cacerts'
Jun 16 21:42:39 debian charon: 00[CFG] loading aa certificates from '/etc/ipsec.d/aacerts'
Jun 16 21:42:39 debian charon: 00[CFG] loading ocsp signer certificates from '/etc/ipsec.d/ocspcerts'
Jun 16 21:42:39 debian charon: 00[CFG] loading attribute certificates from '/etc/ipsec.d/acerts'
Jun 16 21:42:39 debian charon: 00[CFG] loading crls from '/etc/ipsec.d/crls'
Jun 16 21:42:39 debian charon: 00[CFG] loading secrets from '/etc/ipsec.secrets'
Jun 16 21:42:39 debian charon: 00[CFG]   loaded IKE secret for %any
Jun 16 21:42:39 debian charon: 00[CFG]   loaded EAP secret for TLS_VORTEX
Jun 16 21:42:39 debian charon: 00[CFG] loaded 0 RADIUS server configurations
Jun 16 21:42:39 debian charon: 00[CFG] HA config misses local/remote address
Jun 16 21:42:39 debian charon: 00[LIB] loaded plugins: charon test-vectors ldap pkcs11 tpm aesni aes rc2 sha2 sha1 md5 mgf1 rdrand random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl gcrypt af-alg fips-prf gmp curve25519 agent chapoly xcbc cmac hmac ctr ccm gcm curl attr kernel-netlink resolve socket-default connmark farp stroke vici updown eap-identity eap-aka eap-md5 eap-gtc eap-mschapv2 eap-radius eap-tls eap-ttls eap-tnc xauth-generic xauth-eap xauth-pam tnc-tnccs dhcp lookip error-notify certexpire led addrblock unity counters
Jun 16 21:42:39 debian charon: 00[LIB] dropped capabilities, running as uid 0, gid 0
Jun 16 21:42:39 debian charon: 00[JOB] spawning 16 worker threads
Jun 16 21:42:39 debian charon: 01[NET] waiting for data on sockets
Jun 16 21:42:39 debian charon: 05[CFG] received stroke: add connection 'vakrangee1234'
Jun 16 21:42:39 debian charon: 05[CFG] conn vakrangee1234
Jun 16 21:42:39 debian charon: 05[CFG]   left=%any
Jun 16 21:42:39 debian charon: 05[CFG]   leftsubnet=0.0.0.0/0
Jun 16 21:42:39 debian charon: 05[CFG]   leftsourceip=%config
Jun 16 21:42:39 debian charon: 05[CFG]   leftauth=psk
Jun 16 21:42:39 debian charon: 05[CFG]   leftauth2=xauth
Jun 16 21:42:39 debian charon: 05[CFG]   right=157.119.230.26
Jun 16 21:42:39 debian charon: 05[CFG]   rightsubnet=10.101.0.0/16
Jun 16 21:42:39 debian charon: 05[CFG]   rightauth=psk
Jun 16 21:42:39 debian charon: 05[CFG]   rightid=%any
Jun 16 21:42:39 debian charon: 05[CFG]   xauth_identity=TLS_VORTEX
Jun 16 21:42:39 debian charon: 05[CFG]   ike=aes128-aes256-sha256-modp1536
Jun 16 21:42:39 debian charon: 05[CFG]   esp=aes128-aes256-sha256-modp1536
Jun 16 21:42:39 debian charon: 05[CFG]   dpddelay=30
Jun 16 21:42:39 debian charon: 05[CFG]   dpdtimeout=150
Jun 16 21:42:39 debian charon: 05[CFG]   dpdaction=2
Jun 16 21:42:39 debian charon: 05[CFG]   sha256_96=no
Jun 16 21:42:39 debian charon: 05[CFG]   mediation=no
Jun 16 21:42:39 debian charon: 05[CFG]   keyexchange=ikev1
Jun 16 21:42:39 debian charon: 05[KNL] 157.119.230.26 is not a local address or the interface is down
Jun 16 21:42:39 debian charon: 05[CFG] added configuration 'vakrangee1234'
Jun 16 21:42:39 debian charon: 06[CFG] received stroke: initiate 'vakrangee1234'
Jun 16 21:42:39 debian charon: 06[KNL] using 192.168.0.2 as address to reach 157.119.230.26/32
Jun 16 21:42:39 debian charon: 06[IKE] queueing ISAKMP_VENDOR task
Jun 16 21:42:39 debian charon: 06[IKE] queueing ISAKMP_CERT_PRE task
Jun 16 21:42:39 debian charon: 06[IKE] queueing MAIN_MODE task
Jun 16 21:42:39 debian charon: 06[IKE] queueing ISAKMP_CERT_POST task
Jun 16 21:42:39 debian charon: 06[IKE] queueing ISAKMP_NATD task
Jun 16 21:42:39 debian charon: 06[IKE] queueing QUICK_MODE task
Jun 16 21:42:39 debian charon: 06[IKE] activating new tasks
Jun 16 21:42:39 debian charon: 06[IKE]   activating ISAKMP_VENDOR task
Jun 16 21:42:39 debian charon: 06[IKE]   activating ISAKMP_CERT_PRE task
Jun 16 21:42:39 debian charon: 06[IKE]   activating MAIN_MODE task
Jun 16 21:42:39 debian charon: 06[IKE]   activating ISAKMP_CERT_POST task
Jun 16 21:42:39 debian charon: 06[IKE]   activating ISAKMP_NATD task
Jun 16 21:42:39 debian charon: 06[IKE] sending XAuth vendor ID
Jun 16 21:42:39 debian charon: 06[IKE] sending DPD vendor ID
Jun 16 21:42:39 debian charon: 06[IKE] sending FRAGMENTATION vendor ID
Jun 16 21:42:39 debian charon: 06[IKE] sending NAT-T (RFC 3947) vendor ID
Jun 16 21:42:39 debian charon: 06[IKE] sending draft-ietf-ipsec-nat-t-ike-02\n vendor ID
Jun 16 21:42:39 debian charon: 06[IKE] initiating Main Mode IKE_SA vakrangee1234[1] to 157.119.230.26
Jun 16 21:42:39 debian charon: 06[IKE] IKE_SA vakrangee1234[1] state change: CREATED => CONNECTING
Jun 16 21:42:39 debian charon: 06[CFG] configured proposals: IKE:AES_CBC_128/AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_1536, IKE:AES_CBC_128/AES_CBC_192/AES_CBC_256/AES_CTR_128/AES_CTR_192/AES_CTR_256/CAMELLIA_CBC_128/CAMELLIA_CBC_192/CAMELLIA_CBC_256/CAMELLIA_CTR_128/CAMELLIA_CTR_192/CAMELLIA_CTR_256/3DES_CBC/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/AES_XCBC_96/AES_CMAC_96/HMAC_SHA1_96/PRF_AES128_XCBC/PRF_AES128_CMAC/PRF_HMAC_SHA2_256/PRF_HMAC_SHA2_384/PRF_HMAC_SHA2_512/PRF_HMAC_SHA1/ECP_256/ECP_384/ECP_521/ECP_256_BP/ECP_384_BP/ECP_512_BP/CURVE_25519/CURVE_448/MODP_3072/MODP_4096/MODP_6144/MODP_8192/MODP_2048, IKE:AES_CCM_16_128/AES_CCM_16_192/AES_CCM_16_256/AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256/CHACHA20_POLY1305/CAMELLIA_CCM_16_128/CAMELLIA_CCM_16_192/CAMELLIA_CCM_16_256/AES_CCM_8_128/AES_CCM_8_192/AES_CCM_8_256/AES_CCM_12_128/AES_CCM_12_192/AES_CCM_12_256/AES_GCM_8_128/AES_GCM_8_192/AES_GCM_8_256/AES_GCM_12_128/AES_GCM_12_192/AES_GCM_12_256/CAMELLIA_CCM_8_128/CAMELLIA_CCM_8_192/CAMELLIA_CCM_8_256/CAMELLIA_CCM_12_128/CAMELLIA_CCM_12_192/CAMELLIA_CCM_12_256/PRF_AES128_XCBC/PRF_AES128_CMAC/PRF_HMAC_SHA2_256/PRF_HMAC_SHA2_384/PRF_HMAC_SHA2_512/PRF_HMAC_SHA1/ECP_256/ECP_384/ECP_521/ECP_256_BP/ECP_384_BP/ECP_512_BP/CURVE_25519/CURVE_448/MODP_3072/MODP_4096/MODP_6144/MODP_8192/MODP_2048
Jun 16 21:42:39 debian charon: 06[ENC] generating ID_PROT request 0 [ SA V V V V V ]
Jun 16 21:42:39 debian charon: 06[NET] sending packet: from 192.168.0.2[500] to 157.119.230.26[500] (252 bytes)
Jun 16 21:42:39 debian charon: 03[NET] sending packet: from 192.168.0.2[500] to 157.119.230.26[500]
Jun 16 21:42:39 debian charon: 01[NET] received packet: from 157.119.230.26[500] to 192.168.0.2[500]
Jun 16 21:42:39 debian charon: 01[NET] waiting for data on sockets
Jun 16 21:42:39 debian charon: 08[NET] received packet: from 157.119.230.26[500] to 192.168.0.2[500] (204 bytes)
Jun 16 21:42:39 debian charon: 08[ENC] parsed ID_PROT response 0 [ SA V V V V V V ]
Jun 16 21:42:39 debian charon: 08[IKE] received NAT-T (RFC 3947) vendor ID
Jun 16 21:42:39 debian charon: 08[IKE] received DPD vendor ID
Jun 16 21:42:39 debian charon: 08[IKE] received XAuth vendor ID
Jun 16 21:42:39 debian charon: 08[ENC] received unknown vendor ID: 82:99:03:17:57:a3:60:82:c6:a6:21:de:00:00:00:00
Jun 16 21:42:39 debian charon: 08[IKE] received FRAGMENTATION vendor ID
Jun 16 21:42:39 debian charon: 08[IKE] received FRAGMENTATION vendor ID
Jun 16 21:42:39 debian charon: 08[CFG] selecting proposal:
Jun 16 21:42:39 debian charon: 08[CFG]   proposal matches
Jun 16 21:42:39 debian charon: 08[CFG] received proposals: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_1536
Jun 16 21:42:39 debian charon: 08[CFG] configured proposals: IKE:AES_CBC_128/AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_1536, IKE:AES_CBC_128/AES_CBC_192/AES_CBC_256/AES_CTR_128/AES_CTR_192/AES_CTR_256/CAMELLIA_CBC_128/CAMELLIA_CBC_192/CAMELLIA_CBC_256/CAMELLIA_CTR_128/CAMELLIA_CTR_192/CAMELLIA_CTR_256/3DES_CBC/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/AES_XCBC_96/AES_CMAC_96/HMAC_SHA1_96/PRF_AES128_XCBC/PRF_AES128_CMAC/PRF_HMAC_SHA2_256/PRF_HMAC_SHA2_384/PRF_HMAC_SHA2_512/PRF_HMAC_SHA1/ECP_256/ECP_384/ECP_521/ECP_256_BP/ECP_384_BP/ECP_512_BP/CURVE_25519/CURVE_448/MODP_3072/MODP_4096/MODP_6144/MODP_8192/MODP_2048, IKE:AES_CCM_16_128/AES_CCM_16_192/AES_CCM_16_256/AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256/CHACHA20_POLY1305/CAMELLIA_CCM_16_128/CAMELLIA_CCM_16_192/CAMELLIA_CCM_16_256/AES_CCM_8_128/AES_CCM_8_192/AES_CCM_8_256/AES_CCM_12_128/AES_CCM_12_192/AES_CCM_12_256/AES_GCM_8_128/AES_GCM_8_192/AES_GCM_8_256/AES_GCM_12_128/AES_GCM_12_192/AES_GCM_12_256/CAMELLIA_CCM_8_128/CAMELLIA_CCM_8_192/CAMELLIA_CCM_8_256/CAMELLIA_CCM_12_128/CAMELLIA_CCM_12_192/CAMELLIA_CCM_12_256/PRF_AES128_XCBC/PRF_AES128_CMAC/PRF_HMAC_SHA2_256/PRF_HMAC_SHA2_384/PRF_HMAC_SHA2_512/PRF_HMAC_SHA1/ECP_256/ECP_384/ECP_521/ECP_256_BP/ECP_384_BP/ECP_512_BP/CURVE_25519/CURVE_448/MODP_3072/MODP_4096/MODP_6144/MODP_8192/MODP_2048
Jun 16 21:42:39 debian charon: 08[CFG] selected proposal: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_1536
Jun 16 21:42:39 debian charon: 08[IKE] reinitiating already active tasks
Jun 16 21:42:39 debian charon: 08[IKE]   ISAKMP_VENDOR task
Jun 16 21:42:39 debian charon: 08[IKE]   MAIN_MODE task
Jun 16 21:42:39 debian charon: 08[ENC] generating ID_PROT request 0 [ KE No NAT-D NAT-D ]
Jun 16 21:42:39 debian charon: 08[NET] sending packet: from 192.168.0.2[500] to 157.119.230.26[500] (332 bytes)
Jun 16 21:42:39 debian charon: 03[NET] sending packet: from 192.168.0.2[500] to 157.119.230.26[500]
Jun 16 21:42:39 debian charon: 01[NET] received packet: from 157.119.230.26[500] to 192.168.0.2[500]
Jun 16 21:42:39 debian charon: 01[NET] waiting for data on sockets
Jun 16 21:42:39 debian charon: 09[NET] received packet: from 157.119.230.26[500] to 192.168.0.2[500] (316 bytes)
Jun 16 21:42:39 debian charon: 09[ENC] parsed ID_PROT response 0 [ KE No NAT-D NAT-D ]
Jun 16 21:42:39 debian charon: 09[IKE] local host is behind NAT, sending keep alives
Jun 16 21:42:39 debian charon: 09[IKE] reinitiating already active tasks
Jun 16 21:42:39 debian charon: 09[IKE]   ISAKMP_VENDOR task
Jun 16 21:42:39 debian charon: 09[IKE]   MAIN_MODE task
Jun 16 21:42:39 debian charon: 09[ENC] generating ID_PROT request 0 [ ID HASH ]
Jun 16 21:42:39 debian charon: 09[NET] sending packet: from 192.168.0.2[4500] to 157.119.230.26[4500] (92 bytes)
Jun 16 21:42:39 debian charon: 03[NET] sending packet: from 192.168.0.2[4500] to 157.119.230.26[4500]
Jun 16 21:42:39 debian charon: 01[NET] received packet: from 157.119.230.26[4500] to 192.168.0.2[4500]
Jun 16 21:42:39 debian charon: 01[NET] waiting for data on sockets
Jun 16 21:42:39 debian charon: 10[NET] received packet: from 157.119.230.26[4500] to 192.168.0.2[4500] (92 bytes)
Jun 16 21:42:39 debian charon: 10[ENC] parsed ID_PROT response 0 [ ID HASH ]
Jun 16 21:42:39 debian charon: 10[IKE] queueing MODE_CONFIG task
Jun 16 21:42:39 debian charon: 10[IKE] activating new tasks
Jun 16 21:42:39 debian charon: 10[IKE] nothing to initiate
Jun 16 21:42:39 debian charon: 01[NET] received packet: from 157.119.230.26[4500] to 192.168.0.2[4500]
Jun 16 21:42:39 debian charon: 01[NET] waiting for data on sockets
Jun 16 21:42:39 debian charon: 11[NET] received packet: from 157.119.230.26[4500] to 192.168.0.2[4500] (92 bytes)
Jun 16 21:42:39 debian charon: 11[ENC] parsed TRANSACTION request 734604540 [ HASH CPRQ(X_TYPE X_USER X_PWD) ]
Jun 16 21:42:39 debian charon: 11[ENC] generating TRANSACTION response 734604540 [ HASH CPRP(X_USER X_PWD) ]
Jun 16 21:42:39 debian charon: 11[NET] sending packet: from 192.168.0.2[4500] to 157.119.230.26[4500] (108 bytes)
Jun 16 21:42:39 debian charon: 03[NET] sending packet: from 192.168.0.2[4500] to 157.119.230.26[4500]
Jun 16 21:42:39 debian charon: 01[NET] received packet: from 157.119.230.26[4500] to 192.168.0.2[4500]
Jun 16 21:42:39 debian charon: 01[NET] waiting for data on sockets
Jun 16 21:42:39 debian charon: 12[NET] received packet: from 157.119.230.26[4500] to 192.168.0.2[4500] (92 bytes)
Jun 16 21:42:39 debian charon: 12[ENC] parsed TRANSACTION request 3188582757 [ HASH CPS(X_STATUS) ]
Jun 16 21:42:39 debian charon: 12[IKE] XAuth authentication of 'TLS_VORTEX' (myself) successful
Jun 16 21:42:39 debian charon: 12[IKE] IKE_SA vakrangee1234[1] established between 192.168.0.2[192.168.0.2]...157.119.230.26[157.119.230.26]
Jun 16 21:42:39 debian charon: 12[IKE] IKE_SA vakrangee1234[1] state change: CONNECTING => ESTABLISHED
Jun 16 21:42:39 debian charon: 12[IKE] scheduling reauthentication in 85423s
Jun 16 21:42:39 debian charon: 12[IKE] maximum IKE_SA lifetime 85963s
Jun 16 21:42:39 debian charon: 12[ENC] generating TRANSACTION response 3188582757 [ HASH CPA(X_STATUS) ]
Jun 16 21:42:39 debian charon: 12[NET] sending packet: from 192.168.0.2[4500] to 157.119.230.26[4500] (92 bytes)
Jun 16 21:42:39 debian charon: 12[IKE] activating new tasks
Jun 16 21:42:39 debian charon: 12[IKE]   activating MODE_CONFIG task
Jun 16 21:42:39 debian charon: 12[ENC] generating TRANSACTION request 197393111 [ HASH CPRQ(ADDR DNS) ]
Jun 16 21:42:39 debian charon: 12[NET] sending packet: from 192.168.0.2[4500] to 157.119.230.26[4500] (92 bytes)
Jun 16 21:42:39 debian charon: 03[NET] sending packet: from 192.168.0.2[4500] to 157.119.230.26[4500]
Jun 16 21:42:39 debian charon: 03[NET] sending packet: from 192.168.0.2[4500] to 157.119.230.26[4500]
Jun 16 21:42:43 debian charon: 05[IKE] sending retransmit 1 of request message ID 197393111, seq 4
Jun 16 21:42:43 debian charon: 05[NET] sending packet: from 192.168.0.2[4500] to 157.119.230.26[4500] (92 bytes)
Jun 16 21:42:43 debian charon: 03[NET] sending packet: from 192.168.0.2[4500] to 157.119.230.26[4500]
Jun 16 21:42:50 debian charon: 12[IKE] sending retransmit 2 of request message ID 197393111, seq 4
Jun 16 21:42:50 debian charon: 12[NET] sending packet: from 192.168.0.2[4500] to 157.119.230.26[4500] (92 bytes)
Jun 16 21:42:50 debian charon: 03[NET] sending packet: from 192.168.0.2[4500] to 157.119.230.26[4500]

**Comment by tobiasbrunner** [comment]
Seems your peer is not supporting mode config (or not configured to), so you can't request a virtual IP (`leftsourceip=%config`) and this

> **options: DHCP over IPsec**

isn't as easy to achieve as an alternative. Can you change the server config so it supports mode config? (Or even better IKEv2?)

**Comment by shankar25121995** [comment]
we cannot change the server config it under vendor control ,its is vendor requirement most of the our machine (ATM) is linux only,vendor need to connect the atm via vpn and assigin the virtual ip address to atm to establish the communication between payment server, can you suggest the any other vpn service support this mode options.kindly help me on this
